lib/puppet/provider/user/windows_adsi.rb

```diff
@@ -20,6 +20,27 @@ def groups=(groups)
     user.set_groups(groups, @resource[:membership] == :minimum)
   end
 
+  def groups_insync?(current, should)
+    return false unless current
+
+    # By comparing account SIDs we don't have to worry about case
+    # sensitivity, or canonicalization of account names.
+
+    # Cannot use munge of the group property to canonicalize @should
+    # since the default array_matching comparison is not commutative
+
+    # dupes automatically weeded out when hashes built
+    current_users = Puppet::Util::Windows::ADSI::Group.name_sid_hash(current)
+    specified_users = Puppet::Util::Windows::ADSI::Group.name_sid_hash(should)
+
+    if @resource[:membership] == :inclusive
+      current_users == specified_users
+    else
+      return true if specified_users.empty?
```

Comment:
This seems unnecessary, and I noticed there's similar logic in the group provider:

```
irb(main):001:0> a = {}
=> {}
irb(main):002:0> b = {:foo => 'bar'}
=> {:foo=>"bar"}
irb(main):003:0> (a.keys.to_a & b.keys.to_a) == a.keys.to_a
=> true
```